### ISSUE TITLE
Min SDK 16 for grantUriPermission?

### ISSUE THREAD

**Issue by dschuermann** [open-keychain/open-keychain#1753]
This is the the outcome of a discussion with @cketti 

Currently we use an approach with UUIDs in https://github.com/open-keychain/open-keychain/blob/master/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/provider/TemporaryFileProvider.java to prevent other apps from reading every temp file and not all temp files. We didn't use ``grantUriPermission`` because it didn't work correctly with ``ACTION_SEND`` and ``ACTION_SEND_MULTIPLE`` (see comments on top of class).

This is actually because it has not been implemented properly on SDK < 16. See comment in https://developer.android.com/reference/android/content/Intent.html#ACTION_SEND

So if we want to use ``grantUriPermission``, we should switch to min sdk 16 and always use  ``setClipData(ClipData)`` with ``FLAG_GRANT_READ_URI_PERMISSION`` for all intents.

Currently we would loose 1,74 % of all users when moving from SDK 15 to 16 according to our developer console statistics. I am in favor of moving to SDK 16 and using ``grantUriPermission``. It's a much cleaner approach in my opinion and uses the OS security.

Feedback?

**Comment by Valodim** [comment]
No objections. Make it so.

**Comment by dschuermann** [comment]
Currently on hold: The vulnerability described in my post at https://www.ibr.cs.tu-bs.de/news/ibr/surreptitious-sharing-2016-04-04.xml could maybe also work with ContentProviders working with content URIs. Thus even after using ``FLAG_GRANT_READ_URI_PERMISSION``, it makes sense to obfuscate the file names by using UUIDs. This has been discussed with Commonsware by email.

**Comment by cketti** [comment]
Is there a legitimate use case where OpenKeychain is handed a URI to its own content provider?

**Comment by dschuermann** [comment]
@cketti You are right, there is no legitimate use case, but this still means we need to check the owner of the file even for "content" URIs. When OpenKeychain receives a URI pointing to its own temporary files it should not open it for encryption.

A little bit far fetched scenario:
OpenKeychain gets URI ``content://org.sufficientlysecure.keychain.tempstorage/panama_leak.pdf`` for encryption together with many more legitimate files for encryption. The user doesn't notice that there is also this file selected and encrypts everything to the pub key of the attacker.

Assumptions: This file has been decrypted before and is still in the temp storage, the attacker knows the file name, and can convince the user to encrypt something to him. I confess this is really far fetched...

**Comment by cketti** [comment]
With content URIs you can simply blacklist the URI authority, e.g. `org.sufficientlysecure.keychain.tempstorage`.

**Comment by dschuermann** [comment]
Haha, I forgot that the hard link trick has no impact here. Yup, that fix is simple by blacklisting.
